src/blop/ax/qserver_agent.py

```diff
@@ -0,0 +1,326 @@
+
+    Notes
+    -----
+    For more complex setups, you can configure the Ax client directly via ``self.ax_client``.
+
+    For complete working examples of creating and using an Agent, see the tutorial
+    documentation, particularly :doc:`/tutorials/qserver-experiment`.
+
+
+    """
+
+    def __init__(
+        self,
+        sensors: Sequence[Sensor],
+        dofs: Sequence[DOF],
+        objectives: Sequence[Objective],
+        evaluation: EvaluationFunction = None,
+        acquisition_plan: str = "acquire",
+        dof_constraints: Sequence[DOFConstraint] = None,
+        qserver_control_addr: str = "tcp://localhost:60615",
+        qserver_info_addr: str = "tcp://localhost:60625",
+        zmq_consumer_ip: str = "localhost",
+        zmq_consumer_port: str = "5578",
+        **kwargs: Any,
+    ):
+        super().__init__(
+            sensors=sensors,
+            dofs=dofs,
+            objectives=objectives,
+            evaluation=evaluation,
+            acquisition_plan=acquisition_plan,
+            dof_constraints=dof_constraints,
+            **kwargs,
+        )
+
+        # Instantiate an object that can communicate with the queueserver
+        self.RM = REManagerAPI(
+            zmq_control_addr=qserver_control_addr, zmq_info_addr=qserver_info_addr
+        )  # To Do, Add arguements to class init
+
+        # Should plans be submitted and automatically started, or not?
+        self._queue_autostart = True
+
+        # Instantiate an object that will listen for start and stop documents and call a method on the stop document
+        self.zmq_consumer = ZMQConsumer(zmq_consumer_ip, zmq_consumer_port, callback=self._stop_doc_callback)
+        self.zmq_consumer.start_zmq_listener_thread()
+
+        # Should we do something when there is a new event?
+        self._listen_to_events = True
+
+        # Should new suggestions be made automatically until all of the trials are complete?
+        self.continuous_suggestion = True
+
+        # Learning parameters
+        self.num_itterations = 30
+        self.n_points = 1
+
+        # Variables used to keep track of the current optimization
+        self.current_itteration = 0
+        self.agent_suggestion_uid = None
+        self.trials = None
+        self.acquisition_finished = False
+        self.optimization_problem = None
+
+    def _stop_doc_callback(self, start_doc, stop_doc):
+        """
+        In here we can decide whether our experiment requested has completed
+
+        If it has completed, we can digest the data from it and move on to the next point.
+        """
+
+        if self._listen_to_events:
+            # Mark the current acquisition as finished
+
+            logger.info("A stop document has been received, evaluating")
+
+            # Evaluate it with the evaluation function
+            outcomes = self.optimization_problem.evaluation_function(uid=self.agent_suggestion_uid, suggestions=self.trials)
```

Comment:
`start_doc` should have a `uid` if you prefer to use that over searching. And this is called when the stop doc is generated so it should exist in the database at this point.